File: cpp/disk.cpp
#include "disk.h"
#include <winioctl.h>
#include <stdexcept>
#include <system_error>
#include <malloc.h> // For _aligned_malloc
#include <span>
#include <cstring> // For memcpy

namespace
{ // Anonymous namespace for internal linkage
    constexpr int SECTOR_SIZE = 512;

    DWORD last_error()
    {
        return GetLastError();
    }

    HANDLE open_handle(const std::string &path, char rw)
    {
        DWORD access = (rw == 'w') ? (GENERIC_WRITE | GENERIC_READ) : GENERIC_READ;
        DWORD creation = (rw == 'w' && path.find("\\\\.\\.") == std::string::npos) ? CREATE_ALWAYS : OPEN_EXISTING;

        return CreateFileA(
            path.c_str(),
            access,
            FILE_SHARE_READ | FILE_SHARE_WRITE,
            nullptr,
            creation,
            FILE_FLAG_NO_BUFFERING | FILE_FLAG_WRITE_THROUGH,
            NULL);
    }

    size_t ioctl_wrapper(
        HANDLE handle,
        DWORD control_code,
        const void *in_buffer,
        size_t in_buffer_size,
        void *out_buffer,
        size_t out_buffer_size)
    {
        DWORD bytes_returned = 0;
        BOOL ok = DeviceIoControl(
            handle,
            control_code,
            const_cast<void *>(in_buffer),
            static_cast<DWORD>(in_buffer_size),
            out_buffer,
            static_cast<DWORD>(out_buffer_size),
            &bytes_returned,
            nullptr);

        if (!ok)
        {
            throw std::system_error(last_error(), std::system_category(), "DeviceIoControl failed");
        }
        return bytes_returned;
    }

    size_t get_drive_geometry(HANDLE drive)
    {
        DISK_GEOMETRY_EX geo = {};
        try
        {
            ioctl_wrapper(drive, IOCTL_DISK_GET_DRIVE_GEOMETRY_EX, nullptr, 0, &geo, sizeof(geo));
            return geo.DiskSize.QuadPart;
        }
        catch (const std::system_error &)
        {
            return 0;
        }
    }

    size_t get_file_size(HANDLE drive)
    {
        LARGE_INTEGER size = {};
        if (GetFileSizeEx(drive, &size) == 0)
        {
            return get_drive_geometry(drive);
        }
        return size.QuadPart;
    }

} // anonymous namespace

namespace disk
{

    std::optional<Disk> Disk::open(const std::string &path, char rw, std::optional<bool> fua)
    {
        HANDLE handle = open_handle(path, rw);
        if (handle == INVALID_HANDLE_VALUE)
        {
            std::cerr << "Can't open file!! '" << path << "'\n";
            return std::nullopt;
        }
        return Disk(path, rw, fua, handle);
    }

    Disk::Disk(const std::string &path, char rw, std::optional<bool> fua, HANDLE handle)
        : path_(path),
          rw_(rw),
          handle_(handle),
          size_(get_file_size(handle)),
          lba_shift_(9),
          write_offset_(0),
          sptdwb_(SCSI_IOCTL_DATA_OUT),
          fua_(fua) {}

    Disk::~Disk()
    {
        close();
    }

    Disk::Disk(Disk &&other) noexcept
        : path_(std::move(other.path_)),
          rw_(other.rw_),
          handle_(other.handle_),
          size_(other.size_),
          lba_shift_(other.lba_shift_),
          write_offset_(other.write_offset_),
          sptdwb_(other.sptdwb_),
          fua_(other.fua_)
    {
        other.handle_ = INVALID_HANDLE_VALUE; // Prevent double-close
    }

    Disk &Disk::operator=(Disk &&other) noexcept
    {
        if (this != &other)
        {
            close();
            path_ = std::move(other.path_);
            rw_ = other.rw_;
            handle_ = other.handle_;
            size_ = other.size_;
            lba_shift_ = other.lba_shift_;
            write_offset_ = other.write_offset_;
            sptdwb_ = other.sptdwb_;
            fua_ = other.fua_;
            other.handle_ = INVALID_HANDLE_VALUE;
        }
        return *this;
    }

    void Disk::close()
    {
        if (handle_ != INVALID_HANDLE_VALUE)
        {
            CloseHandle(handle_);
            handle_ = INVALID_HANDLE_VALUE;
        }
    }

    size_t Disk::get_size() const { return size_; }
    HANDLE Disk::get_handle() const { return handle_; }
    const std::string &Disk::get_path() const { return path_; }

    void Disk::scsi_open(const std::string &path)
    {
        close();
        handle_ = open_handle(path, rw_);
        if (handle_ == INVALID_HANDLE_VALUE)
        {
            throw std::system_error(last_error(), std::system_category(), "Failed to open SCSI device");
        }
    }

    size_t Disk::scsi_pass_through_direct()
    {
        return ioctl_wrapper(handle_, IOCTL_SCSI_PASS_THROUGH_DIRECT, &sptdwb_, sizeof(sptdwb_), &sptdwb_, sizeof(sptdwb_));
    }

    uint8_t Disk::get_scsi_address()
    {
        SCSI_ADDRESS scsi_addr = {};
        ioctl_wrapper(handle_, IOCTL_SCSI_GET_ADDRESS, nullptr, 0, &scsi_addr, sizeof(scsi_addr));
        std::cout << "get_scsi_address: port " << scsi_addr.PortNumber << " bus " << scsi_addr.PathId << std::endl;
        return scsi_addr.PathId;
    }

    void Disk::get_cache_information()
    {
        DISK_CACHE_INFORMATION cache_info = {};
        ioctl_wrapper(handle_, IOCTL_DISK_GET_CACHE_INFORMATION, nullptr, 0, &cache_info, sizeof(cache_info));
        std::cout << "get_cache_information: WriteCacheEnabled = " << (cache_info.WriteCacheEnabled ? "true" : "false") << std::endl;
    }

    void Disk::storage_query_property()
    {
        STORAGE_PROPERTY_QUERY spq = {};
        spq.PropertyId = StorageDeviceWriteCacheProperty;
        spq.QueryType = PropertyStandardQuery;
        STORAGE_WRITE_CACHE_PROPERTY cache_prop = {};
        ioctl_wrapper(handle_, IOCTL_STORAGE_QUERY_PROPERTY, &spq, sizeof(spq), &cache_prop, sizeof(cache_prop));
        std::cout << "storage_query_property: WriteCacheEnabled = " << static_cast<int>(cache_prop.WriteCacheEnabled) << std::endl;
    }

    size_t Disk::security_recv(uint8_t protocol, uint16_t com_id, std::span<uint8_t> buf)
    {
        scsi::ScsiSecCdb12 cdb(scsi::ScsiOpcode::SECURITY_RECV, protocol, com_id, static_cast<uint32_t>(buf.size()));
        sptdwb_.set_buffer(SCSI_IOCTL_DATA_IN, buf.data(), buf.size());
        sptdwb_.sptd.CdbLength = 12;
        memcpy(sptdwb_.sptd.Cdb, &cdb, sizeof(cdb));

        return scsi_pass_through_direct();
    }

    size_t Disk::security_send(uint8_t protocol, uint16_t com_id, std::span<const uint8_t> buf)
    {
        scsi::ScsiSecCdb12 cdb(scsi::ScsiOpcode::SECURITY_SEND, protocol, com_id, static_cast<uint32_t>(buf.size()));
        sptdwb_.set_buffer(SCSI_IOCTL_DATA_OUT, std::span<const uint8_t>(buf));
        sptdwb_.sptd.CdbLength = 12;
        memcpy(sptdwb_.sptd.Cdb, &cdb, sizeof(cdb));

        return scsi_pass_through_direct();
    }

    size_t Disk::discovery0()
    {
        std::vector<uint8_t> buff(4096, 0);
        auto res = security_recv(0x01, 0x0001, buff);
        // Maybe print just a portion
        std::cout << "discovery0 received " << res << " bytes" << std::endl;
        return res;
    }

    void Disk::storage_set_property()
    {
        // Caution: This is a complex operation. The Rust code is also incomplete.
        // This is a placeholder showing the IOCTL call.
        STORAGE_PROPERTY_SET sps = {};
        sps.PropertyId = StorageDeviceWriteCacheProperty;
        sps.SetType = PropertyStandardSet;
        // The buffer to set follows the sps structure in memory.
        // This is a simplified and likely incorrect version.
        char buffer[sizeof(STORAGE_PROPERTY_SET) + sizeof(STORAGE_WRITE_CACHE_PROPERTY)] = {};
        memcpy(buffer, &sps, sizeof(sps));

        STORAGE_WRITE_CACHE_PROPERTY cache_prop_to_set = {}; // Set properties here
        memcpy(buffer + sizeof(STORAGE_PROPERTY_SET), &cache_prop_to_set, sizeof(cache_prop_to_set));

        ioctl_wrapper(handle_, IOCTL_STORAGE_SET_PROPERTY, buffer, sizeof(buffer), nullptr, 0);
        std::cout << "storage_set_property called." << std::endl;
    }

    size_t Disk::scsi_read(uint64_t offset, std::span<uint8_t> buf)
    {
        if (buf.empty())
            return 0;

        size_t len = (buf.size() + SECTOR_SIZE - 1) & ~(SECTOR_SIZE - 1);
        if (len > buf.size())
            len = buf.size(); // Do not overflow buffer

        uint64_t lba = offset >> lba_shift_;
        uint32_t nlb = static_cast<uint32_t>(len >> lba_shift_);

        scsi::ScsiRwCdb16 cdb(scsi::ScsiOpcode::READ_16, lba, nlb, 0);
        sptdwb_.set_buffer(SCSI_IOCTL_DATA_IN, buf.data(), len);
        sptdwb_.sptd.CdbLength = 16;
        memcpy(sptdwb_.sptd.Cdb, &cdb, sizeof(cdb));

        return scsi_pass_through_direct();
    }

    size_t Disk::scsi_write(std::span<const uint8_t> buf)
    {
        if (buf.empty())
            return 0;
        size_t len = (buf.size() + SECTOR_SIZE - 1) & ~(SECTOR_SIZE - 1);
        if (len > buf.size())
            len = buf.size();

        uint64_t lba = write_offset_ >> lba_shift_;
        uint32_t nlb = static_cast<uint32_t>(len >> lba_shift_);

        uint8_t flag = 0;
        if (fua_.has_value() && fua_.value())
        {
            flag = static_cast<uint8_t>(scsi::ScsiCdbFlag::FUA);
        }

        scsi::ScsiRwCdb16 cdb(scsi::ScsiOpcode::WRITE_16, lba, nlb, flag);
        sptdwb_.set_buffer(SCSI_IOCTL_DATA_OUT, const_cast<uint8_t *>(buf.data()), len);
        sptdwb_.sptd.CdbLength = 16;
        memcpy(sptdwb_.sptd.Cdb, &cdb, sizeof(cdb));

        scsi_pass_through_direct();
        write_offset_ += sptdwb_.sptd.DataTransferLength;
        return sptdwb_.sptd.DataTransferLength;
    }

    size_t Disk::read(std::span<uint8_t> buf)
    {
        if (buf.empty())
        {
            return 0;
        }
        size_t len = (buf.size() + SECTOR_SIZE - 1) & ~(SECTOR_SIZE - 1);
        void *aligned_buf = _aligned_malloc(len, SECTOR_SIZE);
        if (!aligned_buf)
        {
            throw std::bad_alloc();
        }

        DWORD bytes_read = 0;
        BOOL res = ReadFile(handle_, aligned_buf, static_cast<DWORD>(len), &bytes_read, nullptr);
        if (!res)
        {
            _aligned_free(aligned_buf);
            throw std::system_error(last_error(), std::system_category(), "ReadFile failed");
        }
        memcpy(buf.data(), aligned_buf, std::min(buf.size(), static_cast<size_t>(bytes_read)));
        _aligned_free(aligned_buf);
        return bytes_read;
    }

    size_t Disk::write(std::span<const uint8_t> buf)
    {
        if (buf.empty())
            return 0;

        size_t len = (buf.size() + SECTOR_SIZE - 1) & ~(SECTOR_SIZE - 1);
        void *aligned_buf = _aligned_malloc(len, SECTOR_SIZE);
        if (!aligned_buf)
        {
            throw std::bad_alloc();
        }
        memset(aligned_buf, 0, len);
        memcpy(aligned_buf, buf.data(), buf.size());

        DWORD bytes_written = 0;
        BOOL res = WriteFile(handle_, aligned_buf, static_cast<DWORD>(len), &bytes_written, nullptr);
        _aligned_free(aligned_buf);
        if (!res)
        {
            throw std::system_error(last_error(), std::system_category(), "WriteFile failed");
        }
        return bytes_written;
    }

    void Disk::flush()
    {
        // The Rust implementation is empty, so this is too.
        // Could call FlushFileBuffers(handle_);
    }

    std::ostream &operator<<(std::ostream &os, const Disk &disk)
    {
        os << "Disk path: \"" << disk.get_path() << "\", handle: " << disk.get_handle()
           << ", size: " << disk.get_size();
        return os;
    }

    int get_physical_drv_number_from_logical_drv(const std::string &drive_name)
    {
        std::string path = "\\\\.\\" + drive_name;
        HANDLE h_device = open_handle(path, 'r');
        if (h_device == INVALID_HANDLE_VALUE)
        {
            return -1;
        }

        int disk_number = -1;
        VOLUME_DISK_EXTENTS extents = {};
        try
        {
            ioctl_wrapper(h_device, IOCTL_VOLUME_GET_VOLUME_DISK_EXTENTS, nullptr, 0, &extents, sizeof(extents));
            disk_number = extents.Extents[0].DiskNumber;
        }
        catch (const std::system_error &e)
        {
            std::cerr << "ioctl_wrapper failed for get_physical_drv_number_from_logical_drv: " << e.what() << std::endl;
        }

        CloseHandle(h_device);
        return disk_number;
    }

} // namespace disk

File: rust/src/dev/disk.rs
use endian_codec::EncodeBE;
use log::warn;
use windows_sys::{
    Win32::Foundation::*, Win32::Storage::FileSystem::*, Win32::Storage::IscsiDisc::*,
    Win32::System::Ioctl::*, Win32::System::IO::*,
};

use std::{
    ffi::c_void,
    fmt,
    io::{self, Read, Write},
    mem::{size_of_val, zeroed},
    ptr::{null, null_mut},
};

use super::scsi::*;
use crate::SECTOR_SIZE;

pub fn last_error() -> u32 {
    unsafe { GetLastError() }
}

pub fn open(path: &str, rw: char) -> isize {
    let filename = std::ffi::CString::new(path).unwrap();
    let handle = unsafe {
        CreateFileA(
            filename.as_ptr() as *const u8,
            if rw == 'w' {
                GENERIC_WRITE | GENERIC_READ
            } else {
                GENERIC_READ
            },
            FILE_SHARE_READ | FILE_SHARE_WRITE,
            null_mut(),
            if rw == 'w' && !path.contains("\\\\.\\") {
                CREATE_ALWAYS
            } else {
                OPEN_EXISTING
            },
            FILE_FLAG_NO_BUFFERING | FILE_FLAG_WRITE_THROUGH,
            0,
        )
    };
    handle
}

pub fn ioctl(
    handle: isize,
    control_code: u32,
    in_buffer: Option<(*const c_void, usize)>,
    out_buffer: Option<(*mut c_void, usize)>,
) -> io::Result<usize> {
    let mut bytes_returned = 0u32;
    let (in_buffer, in_buffer_size) = in_buffer.unwrap_or((null(), 0));
    let (out_buffer, out_buffer_size) = out_buffer.unwrap_or((null_mut(), 0));
    let ok = unsafe {
        DeviceIoControl(
            handle,
            control_code,
            in_buffer,
            in_buffer_size as u32,
            out_buffer,
            out_buffer_size as u32,
            &mut bytes_returned,
            null_mut(),
        )
    };
    if ok == 0 {
        Err(io::Error::new(
            io::ErrorKind::Other,
            format!("Error code: {:#08x}", last_error()),
        ))
    } else {
        Ok(bytes_returned as usize)
    }
}

fn geometry(drive: &HANDLE) -> usize {
    let mut geo: DISK_GEOMETRY_EX = unsafe { zeroed() };
    if let Ok(_r) = {
        ioctl(
            *drive,
            IOCTL_DISK_GET_DRIVE_GEOMETRY_EX,
            Some((null_mut(), 0)),
            Some((&mut geo as *mut _ as *mut c_void, size_of_val(&geo))),
        )
    } {
        geo.DiskSize as usize
    } else {
        0 as usize
    }
}

fn getfilesize(drive: &HANDLE) -> usize {
    let mut bytes_returned = 0;
    let r = unsafe { GetFileSizeEx(*drive, &mut bytes_returned) };
    if r == 0 {
        geometry(drive)
    } else {
        bytes_returned as usize
    }
}

pub struct Disk {
    path: String,
    rw: char,
    pub handle: HANDLE,
    pub size: usize,
    pub lba_shift: u8,
    write_offset: u64,
    sptdwb: SCSI_PASS_THROUGH_DIRECT_WITH_BUFFER,
    pub fua: Option<bool>,
}

impl fmt::Debug for Disk {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        fmt::Display::fmt(self, fmt)
    }
}

impl fmt::Display for Disk {
    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
        write!(
            fmt,
            "Disk path: \"{}\", rw: '{}', handle: {}, size: {}, fua: {:?}",
            self.path, self.rw, self.handle, self.size, self.fua,
        )
    }
}

impl Disk {
    pub fn open(path: String, rw: char, fua: Option<bool>) -> Option<Disk> {
        let handle = open(&path, rw);
        if handle == INVALID_HANDLE_VALUE {
            warn!("Can't open file!! '{}'", path);
            None
        } else {
            Some(Disk {
                path,
                rw,
                handle,
                size: getfilesize(&handle),
                lba_shift: 9,
                write_offset: 0,
                sptdwb: SCSI_PASS_THROUGH_DIRECT_WITH_BUFFER::new(SCSI_IOCTL_DATA_OUT),
                fua,
            })
        }
    }

    /// The size of the drive in bytes.
    pub fn size(&self) -> usize {
        self.size
    }

    pub fn scsi_open(&mut self, path: String) {
        unsafe { CloseHandle(self.handle) };
        self.handle = INVALID_HANDLE_VALUE;

        let handle = open(&path, self.rw);
        if handle == INVALID_HANDLE_VALUE {
            warn!("Can't open file!! '{}'", path);
        } else {
            self.handle = handle;
        }
    }

    pub fn get_scsi_address(&self) -> io::Result<u8> {
        let mut scsi_addr: SCSI_ADDRESS = unsafe { zeroed() };
        if let Ok(_r) = ioctl(
            self.handle,
            IOCTL_SCSI_GET_ADDRESS,
            Some((null_mut(), 0)),
            Some((
                &mut scsi_addr as *mut _ as *mut c_void,
                size_of_val(&scsi_addr),
            )),
        ) {
            println!(
                "get_scsi_address: port {} bus {}",
                scsi_addr.PortNumber, scsi_addr.PathId
            );
            Ok(scsi_addr.PathId)
        } else {
            Err(io::Error::new(
                io::ErrorKind::Other,
                format!("Error code: {:#08x}", last_error()),
            ))
        }
    }

    pub fn get_cache_information(&self) {
        let mut cache_info: DISK_CACHE_INFORMATION = unsafe { zeroed() };
        if let Ok(_r) = ioctl(
            self.handle,
            IOCTL_DISK_GET_CACHE_INFORMATION,
            Some((null_mut(), 0)),
            Some((
                &mut cache_info as *mut _ as *mut c_void,
                size_of_val(&cache_info),
            )),
        ) {
            println!("get_cache_information: {:?}", cache_info.WriteCacheEnabled);
        }
    }

    pub fn storage_query_property(&self) {
        let mut spq: STORAGE_PROPERTY_QUERY = unsafe { zeroed() };
        spq.PropertyId = StorageDeviceWriteCacheProperty;
        let mut cache_info: STORAGE_WRITE_CACHE_PROPERTY = unsafe { zeroed() };
        if let Ok(_r) = ioctl(
            self.handle,
            IOCTL_STORAGE_QUERY_PROPERTY,
            Some((&spq as *const _ as *const c_void, size_of_val(&spq))),
            Some((
                &mut cache_info as *mut _ as *mut c_void,
                size_of_val(&cache_info),
            )),
        ) {
            println!("storage_query_property: {}", cache_info.WriteCacheEnabled);
        }
    }

    pub fn scsi_pass_through_direct(&mut self) -> io::Result<usize> {
        ioctl(
            self.handle,
            IOCTL_SCSI_PASS_THROUGH_DIRECT,
            Some((
                &self.sptdwb as *const _ as *const c_void,
                size_of_val(&self.sptdwb),
            )),
            Some((
                &mut self.sptdwb as *mut _ as *mut c_void,
                size_of_val(&self.sptdwb),
            )),
        )
    }

    pub fn security_recv(&mut self, protocol: u8, com_id: u16, buf: &[u8]) -> io::Result<usize> {
        let cdb = ScsiSecCdb12::new(
            ScsiOpcode::SCSI_OPCODE_SECURITY_RECV,
            protocol,
            com_id,
            buf.len() as u32,
        );
        self.sptdwb.set_buffer(SCSI_IOCTL_DATA_IN, buf);
        cdb.encode_as_be_bytes(&mut self.sptdwb.sptd.Cdb);
        self.sptdwb.sptd.CdbLength = 12;

        self.scsi_pass_through_direct()
    }

    pub fn security_send(&mut self, protocol: u8, com_id: u16, buf: &[u8]) -> io::Result<usize> {
        let cdb = ScsiSecCdb12::new(
            ScsiOpcode::SCSI_OPCODE_SECURITY_SEND,
            protocol,
            com_id,
            buf.len() as u32,
        );
        self.sptdwb.set_buffer(SCSI_IOCTL_DATA_OUT, buf);
        cdb.encode_as_be_bytes(&mut self.sptdwb.sptd.Cdb);
        self.sptdwb.sptd.CdbLength = 12;

        self.scsi_pass_through_direct()
    }

    pub fn discovery0(&mut self) -> io::Result<usize> {
        let buff = Box::new(vec![0x0u8; 4096]);
        let res = self.security_recv(0x01, 0x0001, &buff);
        println!("discovery0 {:?}", &buff[..512]);
        res
    }

    pub fn storage_set_property(&self) {
        let mut sps: STORAGE_PROPERTY_SET = unsafe { zeroed() };
        sps.PropertyId = StorageDeviceWriteCacheProperty;
        sps.SetType = PropertyStandardSet;
        let mut cache_info: STORAGE_WRITE_CACHE_PROPERTY = unsafe { zeroed() };

        if let Ok(_r) = ioctl(
            self.handle,
            IOCTL_STORAGE_SET_PROPERTY,
            Some((&sps as *const _ as *const c_void, size_of_val(&sps))),
            Some((
                &mut cache_info as *mut _ as *mut c_void,
                size_of_val(&cache_info),
            )),
        ) {
            println!("storage_set_property: {}", cache_info.WriteCacheEnabled);
        }
    }

    pub fn scsi_read(&mut self, offset: u64, buf: &[u8]) -> io::Result<usize> {
        if buf.len() <= 0 {
            return Ok(0);
        }
        let mut len = buf.len() - 1;
        len += SECTOR_SIZE - (len % SECTOR_SIZE);
        let lba = offset >> self.lba_shift;
        let nlb = (len as u32 >> self.lba_shift) - 1;
        let cdb = ScsiRwCdb16::new(ScsiOpcode::SCSI_OPCODE_READ_16, lba, nlb, 0);
        self.sptdwb.set_buffer(SCSI_IOCTL_DATA_IN, buf);
        cdb.encode_as_be_bytes(&mut self.sptdwb.sptd.Cdb);
        self.sptdwb.sptd.CdbLength = 16;

        self.scsi_pass_through_direct()
    }

    #[allow(unused_assignments)]
    pub fn scsi_write(&mut self, buf: &[u8]) -> io::Result<usize> {
        if buf.len() <= 0 {
            return Ok(0);
        }
        let mut len = buf.len() - 1;
        len += SECTOR_SIZE - (len % SECTOR_SIZE);
        let lba = self.write_offset >> self.lba_shift;
        let nlb = (len as u32 >> self.lba_shift) - 1;
        let mut flag = 0;
        if let Some(fua) = self.fua {
            flag = (fua as u8) << 3;
        }
        let cdb = ScsiRwCdb16::new(ScsiOpcode::SCSI_OPCODE_WRITE_16, lba, nlb, flag as u8);
        self.sptdwb.set_buffer(SCSI_IOCTL_DATA_OUT, buf);
        cdb.encode_as_be_bytes(&mut self.sptdwb.sptd.Cdb);
        self.sptdwb.sptd.CdbLength = 16;

        let res = self.scsi_pass_through_direct();
        match res {
            Err(_err) => Err(io::Error::new(
                io::ErrorKind::Other,
                format!("Error code: {:#08x}", last_error()),
            )),
            Ok(_wb) => {
                self.write_offset += self.sptdwb.sptd.DataTransferLength as u64;
                Ok(self.sptdwb.sptd.DataTransferLength as usize)
            }
        }
    }
}

impl Read for Disk {
    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
        let mut bytes_read = 0u32;
        let res = unsafe {
            ReadFile(
                self.handle,
                buf.as_mut_ptr() as *mut u8,
                buf.len() as u32,
                &mut bytes_read,
                null_mut(),
            )
        };
        if res == 0 {
            Err(io::Error::new(
                io::ErrorKind::Other,
                format!("Error code: {:#08x}", last_error()),
            ))
        } else {
            Ok(bytes_read as usize)
        }
    }
}

impl Write for Disk {
    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
        if buf.len() <= 0 {
            return Ok(0);
        }
        let mut len = buf.len() - 1;
        len += SECTOR_SIZE - (len % SECTOR_SIZE);
        let mut bytes_write = 0u32;
        let res = unsafe {
            WriteFile(
                self.handle,
                buf.as_ptr() as *const u8,
                len as u32,
                &mut bytes_write,
                null_mut(),
            )
        };
        if res == 0 {
            Err(io::Error::new(
                io::ErrorKind::Other,
                format!("Error code: {:#08x}", last_error()),
            ))
        } else {
            Ok(bytes_write as usize)
        }
    }

    fn flush(&mut self) -> io::Result<()> {
        Ok(())
    }
}

unsafe impl Send for Disk {}
unsafe impl Sync for Disk {}

pub fn get_physical_drv_number_from_logical_drv(drive_name: String) -> i32 {
    let mut disk_number = -1;
    let path = format!("\\\\.\\{drive_name}");
    let h_device = open(&path, 'r');

    if h_device != INVALID_HANDLE_VALUE {
        let mut st_volume_data: VOLUME_DISK_EXTENTS = unsafe { zeroed() };
        if let Ok(_ret) = ioctl(
            h_device,
            IOCTL_VOLUME_GET_VOLUME_DISK_EXTENTS,
            Some((null_mut(), 0)),
            Some((
                &mut st_volume_data as *mut _ as *mut c_void,
                size_of_val(&st_volume_data),
            )),
        ) {
            disk_number = st_volume_data.Extents[0].DiskNumber as i32;
        }
        unsafe { CloseHandle(h_device) };
    }
    disk_number
}
